Ensure appropriate checks on subview specific View constructor

tcclevenger opened the pull request:
Addresses https://github.com/kokkos/kokkos/issues/7735 for `Kokkos_ENABLE_IMPL_VIEW_LEGACY=ON`. For MDSpan impl, we still get an erroneous build error.

Internally in the subview specification for `ViewMapping`, map subview to `traits_type` (subview type presumed by `ViewMapping`), then map that to the `DstTraits` requested by the `View` constructor. This allows for all proper checks to be run on the `DstTraits` type to ensure no illegal subview types are being requested by the user.

Previously, we blindly mapped to `DstTraits` with no layout checking, which resulted in strided subviews being assigned to non-stided layouts. Also, there were cases where subviews were not strided, but internal build errors stopped appropriate types from being assigned (e.g., LayoutLeft 2d view, subviewed along dim0 (contig.), did not allow assignment to LayoutRight 1d view).

Added a test for the situations above.

Note: We cannot simply deprecate the subview constructor as suggested in the issue, since this is what `Kokkos::subview()` uses to return the subview.

tcclevenger committed:
Fix bug allowing incorrect subview constructor types

Previously, one could request incorrect view types using the subview
constructor by having the subview type have the same layout as the view
type being subviewed. This lead to situations where subview should be
strided, but was not.

Also, corrects some situations where subview type was too strict, e.g.,
a subview along the first dimension of a 2d, layout left view should
allow for the subview type to be layout right, since memis contiguous.

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger committed:
Add test for subview construction

This test fails with current develop, but is fixed in this PR

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger committed:
Comment out part of test which doesn't compile for BasicView

I think this line should compile?

Signed-off-by: tcclevenger <tccleve@sandia.gov>

masterleinad commented:
> Note: We cannot simply deprecate the subview constructor as suggested in the issue, since this is what `Kokkos::subview()` uses to return the subview.

I was thinking about enabling something like
```diff
diff --git a/core/src/Kokkos_View.hpp b/core/src/Kokkos_View.hpp
index 4987a940a..b280d27d8 100644
--- a/core/src/Kokkos_View.hpp
+++ b/core/src/Kokkos_View.hpp
@@ -34,6 +34,11 @@ static_assert(false,
 #include <View/Kokkos_ViewTraits.hpp>
 #include <Kokkos_MemoryTraits.hpp>
 
+namespace Kokkos::Impl {
+template <class DstType, class SrcType, class... Args>
+struct CommonSubview;
+}
+
 // FIXME: This will eventually be removed
 namespace Kokkos::Impl {
 template <class, class...>
@@ -558,11 +563,20 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
   // Compatible subview constructor
   // may assign unmanaged from managed.
 
+private:
   template <class RT, class... RP, class Arg0, class... Args>
   KOKKOS_INLINE_FUNCTION View(const View<RT, RP...>& src_view, const Arg0 arg0,
                               Args... args)
       : base_t(Impl::subview_ctor_tag, src_view, arg0, args...) {}
 
+template <class D, class... P, class... Args>
+friend KOKKOS_INLINE_FUNCTION auto subview(const View<D, P...>& src, Args... args);
+
+template <class DstType, class SrcType, class... Args>
+friend struct ::Kokkos::Impl::CommonSubview;
+
+public:
+
   //----------------------------------------
   // Allocation according to allocation properties and array layout
 ```
when deprecated code is disabled.

crtrott commented on core/unit_test/TestSubView_c15.hpp:
```diff
@@ -0,0 +1,28 @@
+
+#ifndef KOKKOS_TEST_SUBVIEW_C15_HPP
+#define KOKKOS_TEST_SUBVIEW_C15_HPP
+#include <TestViewSubview.hpp>
```
I don't think we need any of the combinatorial machinery in the header file, so you can just include that standalone test (test_subview_constructor_types) directly here or?

crtrott commented on core/src/View/Kokkos_ViewLegacy.hpp:
```diff
@@ -950,18 +950,9 @@ class View : public ViewTraits<DataType, Properties...> {
-        "Subview construction requires compatible view and subview arguments");
-
-    Mapping::assign(m_map, src_view.m_map, arg0, args...);
+    Mapping::assign(m_map, src_view.m_map, m_track.m_tracker, arg0, args...);
```
I don't get why we need the tracker where we didn't before. It also means that we assign tracker twice? Doesn't it do that up on top?

tcclevenger replied:
I needed to pass it to the next `ViewMapping<Dst, Src>::assign()` which passes it to `ViewDataHandle::assign()`. But, no overload of `ViewDataHandle` in the entire code base actually does anything with this tracker (that I can see).

I've changed it to pass a dummy tracker instead.

crtrott reviewed (changes requested):
I don't think you need to follow the structure of the other subview tests here, since you are not reusing any of the facilities in the TestSubView header. Also not sure what's up with the tracker thingy.

crtrott commented:
Oh but thanks for working on this. I do think we should fix this even if the new impl doesn't have the issue.

crtrott commented on core/unit_test/TestViewSubview.hpp:
```diff
@@ -2320,6 +2320,30 @@ struct TestExtentsStaticTests {
+    // any view layout is appropriate.
+    (void)Kokkos::View<int*, LL>(a, Kokkos::ALL, 1);
+    (void)Kokkos::View<int*, LS>(a, Kokkos::ALL, 1);
+    //(void)Kokkos::View<int*, LR>(a, Kokkos::ALL, 1); // FIXME: This doesn't compile for BasicView?
```
@nmm0 you wanna take a look why this doesn't work with BasicView?

tcclevenger committed:
Move subview test out of TestViewSubview.hpp

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger committed:
Remove tracker from assign()

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger committed:
Fix formatting

Signed-off-by: tcclevenger <tccleve@sandia.gov>

dalg24 commented on core/unit_test/TestSubView_c15.hpp:
```diff
@@ -0,0 +1,51 @@
+//
+//@HEADER
+
+#ifndef KOKKOS_TEST_SUBVIEW_C15_HPP
+#define KOKKOS_TEST_SUBVIEW_C15_HPP
```
```suggestion
```
Header guards are unwarranted here

dalg24 commented on core/unit_test/TestSubView_c15.hpp:
```diff
@@ -0,0 +1,51 @@
+#include <gtest/gtest.h>
+#include <Kokkos_Core.hpp>
+
+namespace Test {
```
```suggestion
namespace {
```

dalg24 commented on core/src/View/Kokkos_ViewMapping.hpp:
```diff
@@ -3313,25 +3313,32 @@ class ViewMapping<
   KOKKOS_INLINE_FUNCTION static void assign(
       ViewMapping<DstTraits, void>& dst,
       ViewMapping<SrcTraits, void> const& src, Args... args) {
-    static_assert(ViewMapping<DstTraits, traits_type, void>::is_assignable,
-                  "Subview destination type must be compatible with subview "
-                  "derived type");
```
Please briefly explain why these compile-time assignability checks are removed.

tcclevenger replied:
I removed because we are mapping to `traits_type` here, instead of to `DstTraits` like before, so the equiv of this check would be an identity map. Then below we call `ViewMapping<DstTraits, traits_type>::assign()`, which itself contains all the important checks for `traits_type` to `DstType`.

dalg24 reviewed (commented):
Agreeing with https://github.com/kokkos/kokkos/pull/8210#issuecomment-3010245374

tcclevenger committed:
Clean up test

Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger commented:
> > Note: We cannot simply deprecate the subview constructor as suggested in the issue, since this is what `Kokkos::subview()` uses to return the subview.
> 
> I was thinking about enabling something like
> 
> ```diff
> diff --git a/core/src/Kokkos_View.hpp b/core/src/Kokkos_View.hpp
> index 4987a940a..b280d27d8 100644
> --- a/core/src/Kokkos_View.hpp
> +++ b/core/src/Kokkos_View.hpp
> @@ -34,6 +34,11 @@ static_assert(false,
>  #include <View/Kokkos_ViewTraits.hpp>
>  #include <Kokkos_MemoryTraits.hpp>
>  
> +namespace Kokkos::Impl {
> +template <class DstType, class SrcType, class... Args>
> +struct CommonSubview;
> +}
> +
>  // FIXME: This will eventually be removed
>  namespace Kokkos::Impl {
>  template <class, class...>
> @@ -558,11 +563,20 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
>    // Compatible subview constructor
>    // may assign unmanaged from managed.
>  
> +private:
>    template <class RT, class... RP, class Arg0, class... Args>
>    KOKKOS_INLINE_FUNCTION View(const View<RT, RP...>& src_view, const Arg0 arg0,
>                                Args... args)
>        : base_t(Impl::subview_ctor_tag, src_view, arg0, args...) {}
>  
> +template <class D, class... P, class... Args>
> +friend KOKKOS_INLINE_FUNCTION auto subview(const View<D, P...>& src, Args... args);
> +
> +template <class DstType, class SrcType, class... Args>
> +friend struct ::Kokkos::Impl::CommonSubview;
> +
> +public:
> +
>    //----------------------------------------
>    // Allocation according to allocation properties and array layout
> ```
> 
> when deprecated code is disabled.

@masterleinad @dalg24 Do you guys think this makes more sense in a separate PR? If not I can add here.

masterleinad commented:
> @masterleinad @dalg24 Do you guys think this makes more sense in a separate PR? If not I can add here.

I think we can discuss deprecation separately. In particular, we have to investigate how disruptive removing that public constructor from the public interface would be.

dalg24 commented on core/unit_test/TestSubView_c15.hpp:
```diff
@@ -0,0 +1,47 @@
+
+namespace {
+
+TEST(TEST_CATEGORY_DEATH, subview_constructor_types) {
```
```suggestion
TEST(TEST_CATEGORY_DEATH, subview_constructor_types) {
  ::testing::FLAGS_gtest_death_test_style = "threadsafe";
```

dalg24 commented on core/unit_test/TestSubView_c15.hpp:
```diff
@@ -0,0 +1,47 @@
+    // Using subview dims (1, ALL). For a LayoutLeft,
+    // resutling subview must be strided.
+    const std::string msg = "View assignment must have compatible layouts";
+    ASSERT_DEATH(((void)Kokkos::View<int*, LL>(a, 1, Kokkos::ALL)), msg);
```
Add a test where `a.extent(0)` was `1`?
I expect that would not yield an error.

tcclevenger replied:
Added. No error for all 3 layouts, as expected.

dalg24 commented on core/unit_test/TestSubView_c15.hpp:
```diff
@@ -0,0 +1,47 @@
+
+namespace {
+
+TEST(TEST_CATEGORY_DEATH, subview_constructor_types) {
```
Can we come up with a better test name?
I don't get `subview_constructor_types`

tcclevenger replied:
I changed to `view_subview_constructor_layout_compatibility`. Is that better? Maybe it's a bit wordy?

dalg24 replied:
I don't have a better idea at the moment and agree it is better

tcclevenger committed:
Update subview constructor test

- Change test name to view_subview_constructor_layout_compatibility
- Add test when a.extent(0)=1
- Add FLAGS_gtest_death_test_style="threadsafe"

Signed-off-by: tcclevenger <tccleve@sandia.gov>

dalg24 commented on core/unit_test/TestSubView_c15.hpp:
```diff
@@ -0,0 +1,58 @@
+
+  Kokkos::View<int**, LL> a2("A2", 1, N);
+  {
+    // Using subview dims (1, ALL), but first dimension is stride 1. Any subview
```
```suggestion
    // Using subview dims (1, ALL), but the first dimension is stride 1. Any subview
```

dalg24 commented on core/unit_test/TestSubView_c15.hpp:
```diff
@@ -0,0 +1,58 @@
+  }
+  {
+    // Using subview dims (1, ALL). For a LayoutLeft,
+    // resutling subview must be strided.
```
```suggestion
    // resulting subview must be strided.
```

dalg24 commented on core/unit_test/TestSubView_c15.hpp:
```diff
@@ -0,0 +1,58 @@
+    (void)Kokkos::View<int*, LL>(a1, Kokkos::ALL, 1);
+    (void)Kokkos::View<int*, LS>(a1, Kokkos::ALL, 1);
+
+    // FIXME: This doesn't compile for BasicView, but should
+    //(void)Kokkos::View<int*, LR>(a, Kokkos::ALL, 1);
```
Let's make sure we track this somewhere.  (Assuming that we merge with the FIXME.)

tcclevenger replied:
I'll add info to the issue linked in the description. Also, updated the description to mention this is fixed only for legacy view impl.

dalg24 reviewed (commented):
I am more or less ok with the test.
Make sure your description correctly states this PR is **partially** addressing the issue butt hat more work is needed.

tcclevenger committed:
Update core/unit_test/TestSubView_c15.hpp

Co-authored-by: Damien L-G <dalg24+github@gmail.com>
Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger committed:
Update core/unit_test/TestSubView_c15.hpp

Co-authored-by: Damien L-G <dalg24+github@gmail.com>
Signed-off-by: tcclevenger <tccleve@sandia.gov>

tcclevenger committed:
Fix formatting

Signed-off-by: tcclevenger <tccleve@sandia.gov>

crtrott commented:
One thing to note is that the actual intent behind the subview constructor is that it avoids an extra view type conversion. And I think that is still true. For example:

``` c++
View<int*, LayotuLeft> a("A", N,M);
View<int*> sub = subview(a, Kokkos::ALL(), 0);
```
Does the subview operation and then a View conversion because the return type is different.

vs.
```c++
View<int*, LayotuLeft> a("A", N,M);
View<int*> sub(a, Kokkos::ALL(), 0);
```

This does no view type conversion. I have to check how true that still is with the current implementation using mdspan stuff but I do think it is.

dalg24 commented:
Make sure you add this in the 4.7 changelog under bug fixes
